{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "601a4a45-c8ba-469e-886b-b38d85ea95fa",
   "metadata": {},
   "source": [
    "# Setup\n",
    "Install and import required packages \n",
    "Extract data if it has not already been extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39952255-4d71-43e2-a9b8-14b2a0423760",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell 1\n",
    "\n",
    "# Uncomment the line below if there are some packages missing\n",
    "#!pip install -r ../requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "383fee75-30fa-464e-b5a5-d9f77557098a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell 2\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import pandas_plink as pp\n",
    "import statsmodels.api as sm\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from functools import reduce\n",
    "from sklearn.metrics import r2_score\n",
    "import random\n",
    "import os\n",
    "import tarfile\n",
    "import gzip\n",
    "from tools import *\n",
    "\n",
    "################################################################\n",
    "# Uncomment the line below if the files required are missing\n",
    "# extract_data()\n",
    "################################################################\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e4e253e-ea13-4c26-a611-d8b5600282eb",
   "metadata": {},
   "source": [
    "# Ensuring correct installation and setup of plink2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5358bb8b-7fca-462a-912f-7c2767b556d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.0.0-a.6 32-bit (20 Oct 2024)              cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "\n",
      "  plink2 <input flag(s)...> [command flag(s)...] [other flag(s)...]\n",
      "  plink2 --help [flag name(s)...]\n",
      "\n",
      "Commands include --rm-dup list, --make-bpgen, --export, --freq, --geno-counts,\n",
      "--sample-counts, --missing, --hardy, --het, --fst, --indep-pairwise,\n",
      "--r2-phased, --sample-diff, --make-king, --king-cutoff, --pmerge, --pgen-diff,\n",
      "--write-samples, --write-snplist, --make-grm-list, --pca, --glm, --adjust-file,\n",
      "--gwas-ssf, --pheno-svd, --clump, --score-list, --variant-score,\n",
      "--genotyping-rate, --pgen-info, --validate, and --zst-decompress.\n",
      "\n",
      "\"plink2 --help | more\" describes all functions.\n"
     ]
    }
   ],
   "source": [
    "# Cell 3\n",
    "\n",
    "\n",
    "\n",
    "# IF RUNNING THIS RESULTS IN THE ERROR MESSAGE:\n",
    "# /bin/bash: ./plink2: Permission denied\n",
    "# UNCOMMENT THE LINE OF CODE BELOW THIS AND RERUN THE CELL THEN RECOMMENT\n",
    "\n",
    "#######################################\n",
    "#!chmod 700 plink2\n",
    "#!chmod 700 plink\n",
    "#######################################\n",
    "\n",
    "# OTHERWISE, SHOULD OUTPUT SOMETHING LIKE \n",
    "\"\"\"\n",
    "PLINK v2.0.0-a.6 32-bit (20 Oct 2024)              cog-genomics.org/plink/2.0/\n",
    "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
    "\n",
    "  plink2 <input flag(s)...> [command flag(s)...] [other flag(s)...]\n",
    "  plink2 --help [flag name(s)...]\n",
    "\n",
    "Commands include --rm-dup list, --make-bpgen, --export, --freq, --geno-counts,\n",
    "--sample-counts, --missing, --hardy, --het, --fst, --indep-pairwise,\n",
    "--r2-phased, --sample-diff, --make-king, --king-cutoff, --pmerge, --pgen-diff,\n",
    "--write-samples, --write-snplist, --make-grm-list, --pca, --glm, --adjust-file,\n",
    "--gwas-ssf, --pheno-svd, --clump, --score-list, --variant-score,\n",
    "--genotyping-rate, --pgen-info, --validate, and --zst-decompress.\n",
    "\n",
    "\"plink2 --help | more\" describes all functions.\n",
    "\n",
    "\"\"\"\n",
    "!./plink2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7b7970-fd47-4f3c-acf9-3c8329c470e5",
   "metadata": {},
   "source": [
    "# Loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1ef81f7a-0c1b-4e59-958e-d56ff50c9ebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell 4\n",
    "phenotype_data = '../data/GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz'\n",
    "\n",
    "# alleles = .bim (snp. came from vcf file (txt file))\n",
    "# samples = .fam (family id)\n",
    "# genotypes = .bed (genotypes). Rows and columns defined by alleles and samples\n",
    "(alleles, samples, genotypes) = pp.read_plink(\"../data/LDREF/1000G.EUR.*\",\n",
    "                             verbose=False)\n",
    "genotypes = pd.DataFrame(genotypes.compute())\n",
    "P = pd.read_csv(phenotype_data, sep='\\t', compression='gzip')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07730f3f-7b00-449a-a6d9-3801651789cf",
   "metadata": {},
   "source": [
    "# Conducting cis-eQTL analyses across all genes and chromosomes in the 1000 Genomes dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea7dfb2a-978d-4b8e-bc1e-018200aa440c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting analysis for chromosome 1\n",
      "Starting cis-eQTL analyses for chromosome 1\n",
      "0/2479 analyses completed\n",
      "100/2479 analyses completed\n",
      "200/2479 analyses completed\n",
      "300/2479 analyses completed\n",
      "400/2479 analyses completed\n",
      "500/2479 analyses completed\n",
      "600/2479 analyses completed\n",
      "700/2479 analyses completed\n",
      "800/2479 analyses completed\n",
      "900/2479 analyses completed\n",
      "1000/2479 analyses completed\n",
      "1100/2479 analyses completed\n",
      "1200/2479 analyses completed\n",
      "1300/2479 analyses completed\n",
      "1400/2479 analyses completed\n",
      "1500/2479 analyses completed\n",
      "1600/2479 analyses completed\n",
      "1700/2479 analyses completed\n",
      "1800/2479 analyses completed\n",
      "1900/2479 analyses completed\n",
      "2000/2479 analyses completed\n",
      "2100/2479 analyses completed\n",
      "2200/2479 analyses completed\n",
      "2300/2479 analyses completed\n",
      "2400/2479 analyses completed\n",
      "2479/2479 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 10\n",
      "Starting cis-eQTL analyses for chromosome 10\n",
      "0/932 analyses completed\n",
      "100/932 analyses completed\n",
      "200/932 analyses completed\n",
      "300/932 analyses completed\n",
      "400/932 analyses completed\n",
      "500/932 analyses completed\n",
      "600/932 analyses completed\n",
      "700/932 analyses completed\n",
      "800/932 analyses completed\n",
      "900/932 analyses completed\n",
      "932/932 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 11\n",
      "Starting cis-eQTL analyses for chromosome 11\n",
      "0/1251 analyses completed\n",
      "100/1251 analyses completed\n",
      "200/1251 analyses completed\n",
      "300/1251 analyses completed\n",
      "400/1251 analyses completed\n",
      "500/1251 analyses completed\n",
      "600/1251 analyses completed\n",
      "700/1251 analyses completed\n",
      "800/1251 analyses completed\n",
      "900/1251 analyses completed\n",
      "1000/1251 analyses completed\n",
      "1100/1251 analyses completed\n",
      "1200/1251 analyses completed\n",
      "1251/1251 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 12\n",
      "Starting cis-eQTL analyses for chromosome 12\n",
      "0/1258 analyses completed\n",
      "100/1258 analyses completed\n",
      "200/1258 analyses completed\n",
      "300/1258 analyses completed\n",
      "400/1258 analyses completed\n",
      "500/1258 analyses completed\n",
      "600/1258 analyses completed\n",
      "700/1258 analyses completed\n",
      "800/1258 analyses completed\n",
      "900/1258 analyses completed\n",
      "1000/1258 analyses completed\n",
      "1100/1258 analyses completed\n",
      "1200/1258 analyses completed\n",
      "1258/1258 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 13\n",
      "Starting cis-eQTL analyses for chromosome 13\n",
      "0/433 analyses completed\n",
      "100/433 analyses completed\n",
      "200/433 analyses completed\n",
      "300/433 analyses completed\n",
      "400/433 analyses completed\n",
      "433/433 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 14\n",
      "Starting cis-eQTL analyses for chromosome 14\n",
      "0/876 analyses completed\n",
      "100/876 analyses completed\n",
      "200/876 analyses completed\n",
      "300/876 analyses completed\n",
      "400/876 analyses completed\n",
      "500/876 analyses completed\n",
      "600/876 analyses completed\n",
      "700/876 analyses completed\n",
      "800/876 analyses completed\n",
      "876/876 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 15\n",
      "Starting cis-eQTL analyses for chromosome 15\n",
      "0/916 analyses completed\n",
      "100/916 analyses completed\n",
      "200/916 analyses completed\n",
      "300/916 analyses completed\n",
      "400/916 analyses completed\n",
      "500/916 analyses completed\n",
      "600/916 analyses completed\n",
      "700/916 analyses completed\n",
      "800/916 analyses completed\n",
      "900/916 analyses completed\n",
      "916/916 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 16\n",
      "Starting cis-eQTL analyses for chromosome 16\n",
      "0/1170 analyses completed\n",
      "100/1170 analyses completed\n",
      "200/1170 analyses completed\n",
      "300/1170 analyses completed\n",
      "400/1170 analyses completed\n",
      "500/1170 analyses completed\n",
      "600/1170 analyses completed\n",
      "700/1170 analyses completed\n",
      "800/1170 analyses completed\n",
      "900/1170 analyses completed\n",
      "1000/1170 analyses completed\n",
      "1100/1170 analyses completed\n",
      "1170/1170 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 17\n",
      "Starting cis-eQTL analyses for chromosome 17\n",
      "0/1242 analyses completed\n",
      "100/1242 analyses completed\n",
      "200/1242 analyses completed\n",
      "300/1242 analyses completed\n",
      "400/1242 analyses completed\n",
      "500/1242 analyses completed\n",
      "600/1242 analyses completed\n",
      "700/1242 analyses completed\n",
      "800/1242 analyses completed\n",
      "900/1242 analyses completed\n",
      "1000/1242 analyses completed\n",
      "1100/1242 analyses completed\n",
      "1200/1242 analyses completed\n",
      "1242/1242 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 18\n",
      "Starting cis-eQTL analyses for chromosome 18\n",
      "0/250 analyses completed\n",
      "100/250 analyses completed\n",
      "200/250 analyses completed\n",
      "250/250 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 19\n",
      "Starting cis-eQTL analyses for chromosome 19\n",
      "0/1265 analyses completed\n",
      "100/1265 analyses completed\n",
      "200/1265 analyses completed\n",
      "300/1265 analyses completed\n",
      "400/1265 analyses completed\n",
      "500/1265 analyses completed\n",
      "600/1265 analyses completed\n",
      "700/1265 analyses completed\n",
      "800/1265 analyses completed\n",
      "900/1265 analyses completed\n",
      "1000/1265 analyses completed\n",
      "1100/1265 analyses completed\n",
      "1200/1265 analyses completed\n",
      "1265/1265 analyses completed\n",
      "Compiling eQTL analysis results\n",
      "Starting analysis for chromosome 2\n",
      "Starting cis-eQTL analyses for chromosome 2\n",
      "0/1617 analyses completed\n"
     ]
    }
   ],
   "source": [
    "# Takes about 1 hour to run. Analysis files have already been created.\n",
    "#all_chromosomes = alleles['chrom'].unique()\n",
    "#output_fp_prefix = 'genome_wide_ciseqtl_analysis/analysis_'\n",
    "#all_outs = []\n",
    "#for chrom in all_chromosomes:\n",
    "#    print(f'Starting analysis for chromosome {chrom}')\n",
    "#    chr_analysis(chrom, alleles, samples, genotypes, P, window=500000).to_csv(output_fp_prefix+chrom+'.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d33b8f7-f39a-4ada-ac76-ab5535a6081c",
   "metadata": {},
   "source": [
    "# Results from the genome wide cis-eQTL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8b5503d7-089e-4c91-8630-495cc7487134",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chr</th>\n",
       "      <th>bp</th>\n",
       "      <th>p</th>\n",
       "      <th>se</th>\n",
       "      <th>beta</th>\n",
       "      <th>a0</th>\n",
       "      <th>a1</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs6426089</td>\n",
       "      <td>1.0</td>\n",
       "      <td>225999979.0</td>\n",
       "      <td>0.030385</td>\n",
       "      <td>2.174076</td>\n",
       "      <td>0.524494</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>86772.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1467143</td>\n",
       "      <td>1.0</td>\n",
       "      <td>226000967.0</td>\n",
       "      <td>0.030385</td>\n",
       "      <td>2.174076</td>\n",
       "      <td>0.524494</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>86773.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs868966</td>\n",
       "      <td>1.0</td>\n",
       "      <td>226006098.0</td>\n",
       "      <td>0.784214</td>\n",
       "      <td>0.274046</td>\n",
       "      <td>0.066733</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>86774.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs10753410</td>\n",
       "      <td>1.0</td>\n",
       "      <td>226008101.0</td>\n",
       "      <td>0.126770</td>\n",
       "      <td>1.530693</td>\n",
       "      <td>0.551515</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>86775.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs10799326</td>\n",
       "      <td>1.0</td>\n",
       "      <td>226009918.0</td>\n",
       "      <td>0.226326</td>\n",
       "      <td>1.212060</td>\n",
       "      <td>0.429218</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>86776.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9234069</th>\n",
       "      <td>rs11998787</td>\n",
       "      <td>9.0</td>\n",
       "      <td>118057870.0</td>\n",
       "      <td>0.932509</td>\n",
       "      <td>-0.084750</td>\n",
       "      <td>-0.001781</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1179771.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9234070</th>\n",
       "      <td>rs9987570</td>\n",
       "      <td>9.0</td>\n",
       "      <td>118061922.0</td>\n",
       "      <td>0.924241</td>\n",
       "      <td>-0.095163</td>\n",
       "      <td>-0.002020</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1179772.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9234071</th>\n",
       "      <td>rs10982660</td>\n",
       "      <td>9.0</td>\n",
       "      <td>118063450.0</td>\n",
       "      <td>0.432366</td>\n",
       "      <td>0.786079</td>\n",
       "      <td>0.012098</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1179773.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9234072</th>\n",
       "      <td>rs10982662</td>\n",
       "      <td>9.0</td>\n",
       "      <td>118064892.0</td>\n",
       "      <td>0.534901</td>\n",
       "      <td>-0.621171</td>\n",
       "      <td>-0.007054</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1179774.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9234073</th>\n",
       "      <td>rs13290457</td>\n",
       "      <td>9.0</td>\n",
       "      <td>118067966.0</td>\n",
       "      <td>0.062737</td>\n",
       "      <td>-1.867141</td>\n",
       "      <td>-0.025066</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1179775.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9234074 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                snp  chr           bp         p        se      beta a0 a1  \\\n",
       "0         rs6426089  1.0  225999979.0  0.030385  2.174076  0.524494  G  A   \n",
       "1         rs1467143  1.0  226000967.0  0.030385  2.174076  0.524494  G  A   \n",
       "2          rs868966  1.0  226006098.0  0.784214  0.274046  0.066733  G  A   \n",
       "3        rs10753410  1.0  226008101.0  0.126770  1.530693  0.551515  C  T   \n",
       "4        rs10799326  1.0  226009918.0  0.226326  1.212060  0.429218  C  T   \n",
       "...             ...  ...          ...       ...       ...       ... .. ..   \n",
       "9234069  rs11998787  9.0  118057870.0  0.932509 -0.084750 -0.001781  A  G   \n",
       "9234070   rs9987570  9.0  118061922.0  0.924241 -0.095163 -0.002020  T  G   \n",
       "9234071  rs10982660  9.0  118063450.0  0.432366  0.786079  0.012098  A  G   \n",
       "9234072  rs10982662  9.0  118064892.0  0.534901 -0.621171 -0.007054  A  G   \n",
       "9234073  rs13290457  9.0  118067966.0  0.062737 -1.867141 -0.025066  A  G   \n",
       "\n",
       "                 i  \n",
       "0          86772.0  \n",
       "1          86773.0  \n",
       "2          86774.0  \n",
       "3          86775.0  \n",
       "4          86776.0  \n",
       "...            ...  \n",
       "9234069  1179771.0  \n",
       "9234070  1179772.0  \n",
       "9234071  1179773.0  \n",
       "9234072  1179774.0  \n",
       "9234073  1179775.0  \n",
       "\n",
       "[9234074 rows x 9 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_chromosomes = alleles['chrom'].unique()\n",
    "output_fp_prefix = 'genome_wide_ciseqtl_analysis/analysis_'\n",
    "all_outs = [output_fp_prefix+chrom+'.txt' for chrom in all_chromosomes]\n",
    "all_chr_ciseqtl_dfs = [pd.read_table(out, index_col='Unnamed: 0') for out in all_outs]\n",
    "genome_wide_ciseqtl_analysis = pd.concat(all_chr_ciseqtl_dfs, ignore_index=True)\n",
    "genome_wide_ciseqtl_analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cd5a070-75a5-4949-a09d-c28b5c02175b",
   "metadata": {},
   "source": [
    "# Getting the significant SNPs from the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a40fe0cf-c77b-42e8-a15c-2f00b0bd29e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p-value: 5.414728103760052e-09\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chr</th>\n",
       "      <th>bp</th>\n",
       "      <th>p</th>\n",
       "      <th>se</th>\n",
       "      <th>beta</th>\n",
       "      <th>a0</th>\n",
       "      <th>a1</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2930</th>\n",
       "      <td>rs9865</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46827456.0</td>\n",
       "      <td>1.826516e-17</td>\n",
       "      <td>-8.981540</td>\n",
       "      <td>-0.361494</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>19475.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2947</th>\n",
       "      <td>rs324419</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46871986.0</td>\n",
       "      <td>4.723186e-13</td>\n",
       "      <td>-7.524525</td>\n",
       "      <td>-0.315464</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>19492.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2955</th>\n",
       "      <td>rs7520850</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46883668.0</td>\n",
       "      <td>8.812540e-17</td>\n",
       "      <td>-8.766008</td>\n",
       "      <td>-0.388400</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>19500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7579</th>\n",
       "      <td>rs10863389</td>\n",
       "      <td>1.0</td>\n",
       "      <td>206656292.0</td>\n",
       "      <td>1.020276e-10</td>\n",
       "      <td>6.671585</td>\n",
       "      <td>0.078233</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>77547.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7581</th>\n",
       "      <td>rs11118092</td>\n",
       "      <td>1.0</td>\n",
       "      <td>206656770.0</td>\n",
       "      <td>8.164971e-14</td>\n",
       "      <td>-7.788729</td>\n",
       "      <td>-0.086488</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>77549.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9221760</th>\n",
       "      <td>rs2805888</td>\n",
       "      <td>9.0</td>\n",
       "      <td>111880736.0</td>\n",
       "      <td>2.903130e-09</td>\n",
       "      <td>6.097941</td>\n",
       "      <td>0.036827</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1176119.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9221761</th>\n",
       "      <td>rs6825</td>\n",
       "      <td>9.0</td>\n",
       "      <td>111881927.0</td>\n",
       "      <td>2.082084e-11</td>\n",
       "      <td>-6.931497</td>\n",
       "      <td>-0.033476</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1176120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9221767</th>\n",
       "      <td>rs12686736</td>\n",
       "      <td>9.0</td>\n",
       "      <td>111888739.0</td>\n",
       "      <td>5.331356e-09</td>\n",
       "      <td>-5.989441</td>\n",
       "      <td>-0.033173</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1176126.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9221769</th>\n",
       "      <td>rs17628095</td>\n",
       "      <td>9.0</td>\n",
       "      <td>111889172.0</td>\n",
       "      <td>5.331356e-09</td>\n",
       "      <td>-5.989441</td>\n",
       "      <td>-0.033173</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1176128.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9221778</th>\n",
       "      <td>rs10759329</td>\n",
       "      <td>9.0</td>\n",
       "      <td>111899928.0</td>\n",
       "      <td>1.158362e-14</td>\n",
       "      <td>8.075820</td>\n",
       "      <td>0.040680</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1176137.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27358 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                snp  chr           bp             p        se      beta a0 a1  \\\n",
       "2930         rs9865  1.0   46827456.0  1.826516e-17 -8.981540 -0.361494  G  A   \n",
       "2947       rs324419  1.0   46871986.0  4.723186e-13 -7.524525 -0.315464  T  C   \n",
       "2955      rs7520850  1.0   46883668.0  8.812540e-17 -8.766008 -0.388400  A  G   \n",
       "7579     rs10863389  1.0  206656292.0  1.020276e-10  6.671585  0.078233  T  C   \n",
       "7581     rs11118092  1.0  206656770.0  8.164971e-14 -7.788729 -0.086488  T  C   \n",
       "...             ...  ...          ...           ...       ...       ... .. ..   \n",
       "9221760   rs2805888  9.0  111880736.0  2.903130e-09  6.097941  0.036827  T  C   \n",
       "9221761      rs6825  9.0  111881927.0  2.082084e-11 -6.931497 -0.033476  C  T   \n",
       "9221767  rs12686736  9.0  111888739.0  5.331356e-09 -5.989441 -0.033173  G  A   \n",
       "9221769  rs17628095  9.0  111889172.0  5.331356e-09 -5.989441 -0.033173  A  G   \n",
       "9221778  rs10759329  9.0  111899928.0  1.158362e-14  8.075820  0.040680  A  G   \n",
       "\n",
       "                 i  \n",
       "2930       19475.0  \n",
       "2947       19492.0  \n",
       "2955       19500.0  \n",
       "7579       77547.0  \n",
       "7581       77549.0  \n",
       "...            ...  \n",
       "9221760  1176119.0  \n",
       "9221761  1176120.0  \n",
       "9221767  1176126.0  \n",
       "9221769  1176128.0  \n",
       "9221778  1176137.0  \n",
       "\n",
       "[27358 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_val = 0.05/genome_wide_ciseqtl_analysis.shape[0]\n",
    "print(f'p-value: {p_val}')\n",
    "genome_wide_ciseqtl_analysis[genome_wide_ciseqtl_analysis['p'] < p_val]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "029290f0-0569-45b6-88f9-290063fa116c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
